### PR TITLE
Update Working with Dependencies -> Artifact Transform page

### PR DESCRIPTION
**Pull request by lkasso**
### Details
This is a Documentation change ONLY.

### Context
For https://github.com/gradle/gradle/issues/27314

### Review
[Link]()

### CODE REVIEW

**Comment by lkasso** [comment]
@bot-gradle test BD

**Comment by lkasso** [comment]
@bot-gradle test BD

**Comment by jvandort** [comment]
@bot-gradle test BD

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 1]
```diff
@@ -108,51 +255,47 @@ include::sample[dir="snippets/dependencyManagement/artifactTransforms-minify/gro
...
 - A relative path.
 
 Gradle uses the absolute path as the location of the output artifact.
-For example, if the input artifact is an exploded WAR, then the transform action can call `TransformOutputs.file()` for all jar files in the `WEB-INF/lib` directory.
+For example, if the input artifact is an exploded WAR, the transform action can call `TransformOutputs.file()` for all JAR files in the `WEB-INF/lib` directory.
 The output of the transform would then be the library JARs of the web application.
 
 For a relative path, the `dir()` or `file()` method returns a workspace to the transform action.
-The implementation of the transform action needs to create the transformed artifact at the location of the provided workspace.
+Implementing the transform action needs to create the transformed artifact at the location of the provided workspace.
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:291]:
```suggestion
The transform action needs to create the transformed artifact at the location of the provided workspace.
```

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 2]
```diff
@@ -212,11 +361,37 @@ include::sample[dir="snippets/dependencyManagement/artifactTransforms-relocate/g
...
+====
+include::sample[dir="snippets/dependencyManagement/artifactTransforms-incremental/kotlin",files="build.gradle.kts[tags=artifact-transform-countloc]"]
+include::sample[dir="snippets/dependencyManagement/artifactTransforms-incremental/groovy",files="build.gradle[tags=artifact-transform-countloc]"]
+====
+<1> Inject `InputChanges`
+<2> Query for changes in the input artifact
+
+// TODO: https://github.com/gradle/gradle/issues/10867
+// [[artifact-resolution-queries]]
+// == Artifact resolution queries
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:388]:
```suggestion
// TODO: https://github.com/gradle/gradle/issues/26007
// [[artifact-views]]
// == Artifact Views
```

ArtifactView != ARQ

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 3]
```diff
@@ -212,11 +361,37 @@ include::sample[dir="snippets/dependencyManagement/artifactTransforms-relocate/g
...
+
+== Registering Artifact Transforms
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:390]: After going through this whole document, I am surprised the registering section is at the end. 
I think moving this section to before we describe _how_ to implement a transform will give the user a much better idea of how they're used.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 4]
```diff
@@ -228,39 +403,44 @@ There are a few points to consider when using `registerTransform()`:
...
-For artifact transforms, only the input artifact is an incremental input, and therefore the transform can only query for changes there.
-In order to use link:{groovyDslPath}/org.gradle.work.InputChanges.html[InputChanges] in the transform action, inject it into the action.
-For more information on how to use link:{groovyDslPath}/org.gradle.work.InputChanges.html[InputChanges], see the corresponding documentation for <<custom_tasks#incremental_tasks,incremental tasks>>.
+On the command line, Gradle runs tasks; not Artifact Transforms: `./gradlew build.`
+So how and when does it run transforms?
 
-Here is an example of an incremental transform that counts the lines of code in Java source files:
+A task can declare an input with attributes.
+When running this task, Gradle may run an Artifact Transform to ensure that the input has the proper attributes.
+Then, it will run the task.
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:426]: This whole paragraph seems a bit imprecise to me. What do you mean by `A task can declare an input with attributes.`? Tasks can declare `FileCollections` as inputs, which are derived from resolvable configurations. Those configurations have attributes and may run artifact transforms during resolution.

Something about "input with attributes" does not sound right to me.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 5]
```diff
@@ -228,39 +403,44 @@ There are a few points to consider when using `registerTransform()`:
...
+
+Gradle will check if this request needs to be executed.
+The execution engine will fulfill the request as efficiently as possible.
+If the request is in the in-memory cache (daemon memory), `UP-TO-DATE`, or in the local/remove cache, Gradle will avoid executing it again.
+This means that one execution of a transform can fulfill many requests.
+Gradle takes full advantage of already fulfilled requests.
+
+TIP: In well-declared builds, project dependencies can be fully discovered during task configuration ahead of task execution scheduling.
+If the project dependency is badly declared (e.g., missing task input), the transform execution will happen inside the task.
+What could be a planned transform node in the execution graph becomes a transform request to the parent task.
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:446]: IMO much of this seems way to nitty gritty into the way the execution engine works. 

In practice I think the user really cares about a few things related to parallelism/caching:
- Artifact transforms can be run in parallel
- Artifact transforms are cached
- If separate resolutions used by different tasks require the same transform to be executed on the same artifacts, the transform results will be cached and shared. 

And then a few things related to _when_ transforms are executed:
- Transforms for project dependencies can be discovered early and can be scheduled to run before the task executes
- Transforms for external dependencies are not discovered early and will be executed as part of task execution
- Improper build dependency declarations can cause project artifact transforms to behave like external artifact transforms, delaying the transform to be executed as the task is executed.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc (hunk 2 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:387]: This is an old TO DO I am leaving in there for now.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 6]
```diff
@@ -12,92 +12,239 @@
...
-The external `guava` dependency has two variants:
+Gradle selects the best chain of transforms based on specific rules:
+- If there is only one chain, it is selected.
+- If one chain is a suffix of another, the more specific chain is selected.
+- The shortest chain is preferred.
+- If multiple chains are equally suitable, the selection fails, and an error is reported.
 
-- `org.gradle.usage=java-runtime,org.gradle.libraryelements=jar,minified=false` and
-- `org.gradle.usage=java-api,org.gradle.libraryelements=jar,minified=false`.
+NOTE: The `artifactType` attribute is special because it is only present on resolved artifacts, not on dependencies. Transforms that only change `artifactType` are not considered during configuration resolution but can be used in an link:{javadocPath}/org/gradle/api/artifacts/ArtifactView.html[`ArtifactView`].
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:215]: I deleted this and will put it back in the Artifact View page.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 7]
```diff
@@ -12,92 +12,239 @@
...
-The `artifactType` attribute is special, since it is only present on resolved artifacts and not on dependencies.
-As a consequence, any transform which is only mutating `artifactType` will never be selected when resolving a configuration with only the `artifactType` as requested attribute.
-It will only be considered when using an link:{javadocPath}/org/gradle/api/artifacts/ArtifactView.html[ArtifactView].
-====
+Gradle allows registering Artifact Transforms.
+When the dependency does not have the requested variant, Gradle will try to find a chain of Artifact Transforms to create it.
+
+== Understanding Artifact Transforms and Attribute-Based Resolution
+
+When Gradle resolves a configuration and a dependency does not have a variant with the requested attributes, it attempts to find a chain of Artifact Transforms to create the desired variant.
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:191]: I kept a lot of the original documentation on this page, which means a lot of duplication. I have gone back and re-ordered material and de-duplicated items. Let me know what you think of this second pass.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 8]
```diff
@@ -12,92 +12,239 @@
...
+For example, if a configuration requests `org.gradle.usage=java-api` and `org.gradle.libraryelements=classes`, Gradle will select the variant of the project dependency that matches these attributes (in this case, the `classes` directory).
 
-Of all the found transform chains, Gradle tries to select the best one:
+=== Artifact Transforms
 
-- If there is only one transform chain, it is selected.
-- If there are two transform chains, and one is a suffix of the other one, it is selected.
-- If there is a shortest transform chain, then it is selected.
-- In all other cases, the selection fails and an error is reported.
+Sometimes, a dependency might not have the exact variant with the requested attributes.
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:181]: Part of the de-duplication effort in second pass.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc (hunk 3 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:389]: I am keeping the same order as the overview for continuity and to make it easy to read.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc (hunk 5 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:445]: I wasn't sure how much we wanted to share in this section based on notes/PRs/discussions. I watered it down, but I'm not sure if it's to your liking yet. Let me know

**Comment by jvandort** [comment]
@bot-gradle test bd

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 9]
```diff
@@ -12,92 +12,239 @@
...
+
+A transform must be registered the same way a task is registered using link:{javadocPath}/org/gradle/api/artifacts/dsl/DependencyHandler.html#registerTransform(java.lang.Class,org.gradle.api.Action)[`dependencies.registerTransform()`]:
+
+Here, our transform is registered with the link:{javadocPath}/org/gradle/api/Project.html#dependencies(groovy.lang.Closure)[`dependencies`] configuration:
+
+[source,kotlin]
+----
+dependencies {
+    registerTransform(MyTransform::class) {
+        from.attribute(ArtifactAttributes.ARTIFACT_FORMAT, "jar")
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:112]: Snippetized.

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 10]
```diff
@@ -12,92 +12,239 @@
...
+It must be annotated with the link:{javadocPath}/org/gradle/api/artifacts/transform/InputArtifact.html[`@InputArtifacts`] annotation.
+
+Then, you implement the link:{javadocPath}/org/gradle/api/artifacts/transform/TransformAction.html#transform(org.gradle.api.artifacts.transform.TransformOutputs)[`transform()`] method from the link:{javadocPath}/org/gradle/api/artifacts/transform/TransformAction.html[`TransformAction`] interface.
+This is where you implement the work the transform should do when triggered.
+The method has the link:{javadocPath}/org/gradle/api/artifacts/transform/TransformOutputs.html[`TransformOutputs`] as an argument that defines what the transform produces.
+
+Here, `MyTransform` is the custom transform action that converts a `jar` artifact to a `transformed-jar` artifact:
+[source,kotlin]
+----
+abstract class MyTransform : TransformAction<TransformParameters.None> {
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:73]: Snippetized.

**Comment by jvandort** [comment]
@bot-gradle test bd

File: platforms/documentation/docs/src/snippets/dependencyManagement/artifactTransforms-simple/kotlin/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    override fun transform(outputs: TransformOutputs) {
+        val inputFile = inputArtifact.get().asFile
+        val outputFile = outputs.file(inputFile.name.replace(".jar", "-transformed.jar"))
+        // Perform transformation logic here
+    }
+}
+// end::artifact-transform-imp[]
+
+// tag::artifact-transform-use[]
+val artifactType = Attribute.of("artifactType", String::class.java)
```
Comment (jvandort) [inline platforms/documentation/docs/src/snippets/dependencyManagement/artifactTransforms-simple/kotlin/build.gradle.kts:21]: We should use `ArtifactTypeDefinition.ARTIFACT_TYPE_ATTRIBUTE` (which is the same thing as this)

File: platforms/documentation/docs/src/snippets/dependencyManagement/artifactTransforms-simple/tests/artifactTransformSimple.sample.conf [hunk 12]
```diff
@@ -0,0 +1,2 @@
+executable: gradle
+args: tasks
```
Comment (jvandort) [inline platforms/documentation/docs/src/snippets/dependencyManagement/artifactTransforms-simple/tests/artifactTransformSimple.sample.conf:2]: We should run something other than `tasks` to verify this works. Printing the tasks out will not resolve the `runtimeClasspath` configuration and therefore will not run the transform. 

I would recommend adding a custom resolve method to do so (you can keep it out of the tag:: so its not visible in the snippet) (below is untested but should be close)

```
depenencies {
    implementation("com.google.guava:guava:33.2.1-jre")
}

repositories {
    mavenCentral()
}

tasks.register("resolve") {
    val files = configurations.runtimeClasspath.incoming.files
    doLast {
        assert(files.map { it.name } == listOf("guava-33.2.1-jre.transformed.jar"))
    }
}
```

File: platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc [hunk 13]
```diff
@@ -228,39 +375,36 @@ There are a few points to consider when using `registerTransform()`:
...
+In well-declared builds, project dependencies can be fully discovered during task configuration ahead of task execution scheduling.
+If the project dependency is badly declared (e.g., missing task input), the transform execution will happen inside the task.
+
+It's important to remember that Artifact Transforms:
+
+- can be run in parallel
+- are cacheable
+- are reusable*
+
+*If separate resolutions used by different tasks require the same transform to be executed on the same artifacts, the transform results will be cached and shared.
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/dep-man/04-modeling-features/artifact_transforms.adoc:410]: Something weird is going on here with the asterisks.

**Review by jvandort — APPROVED** [review]

**Pull request merged**